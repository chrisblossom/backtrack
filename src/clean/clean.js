/* @flow */

import del from 'del';
import { makeDirs } from '../utils/make-dirs';

import { buildPath } from '../config/paths';

import type { NormalizedClean } from '../types.js';

const base: NormalizedClean = {
    del: [],
    makeDirs: [],
};

async function clean(args?: NormalizedClean = base): Promise<NormalizedClean> {
    /**
     * Run all del requests together.
     *
     * Ensures you can overwrite presets to exclude files with a glob
     * and all makeDirs are not deleted
     */
    // $FlowIssue
    const delResults = await del(args.del, {
        cwd: buildPath,
        dot: true,
    });

    /**
     * run makeDirs after del has been completed
     */
    await makeDirs(args.makeDirs);

    return {
        del: delResults,
        makeDirs: args.makeDirs,
    };
}

export { clean };
