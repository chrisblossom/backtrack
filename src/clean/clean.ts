import del from 'del';
import { makeDirs } from '../utils/make-dirs';
import { copy } from '../utils/copy';

import { buildPath } from '../config/paths';

import { NormalizedClean } from '../types';

const base: NormalizedClean = {
    del: [],
    makeDirs: [],
    copy: [],
};

async function clean(args: NormalizedClean = base) {
    /**
     * Run all del requests together.
     *
     * Ensures you can overwrite presets to exclude files with a glob
     * and all makeDirs are not deleted
     */
    await del(args.del, {
        cwd: buildPath,
        dot: true,
    });

    /**
     * run makeDirs after del has been completed
     */
    await makeDirs(args.makeDirs);

    /**
     * Copy static files last
     */
    await copy(args.copy);
}

export { clean };
