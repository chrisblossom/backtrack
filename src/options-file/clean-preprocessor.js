/* @flow */

import { toArray } from '../utils/object-utils';
import type { Clean, NormalizedClean } from '../types.js';

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value?: Clean | $ReadOnlyArray<Clean>,
};

function cleanPreprocessor({ value }: Args = {}): $ReadOnlyArray<
    NormalizedClean,
> {
    const allToArray = toArray(value);

    return allToArray.map((arg: Clean) => {
        return {
            del: toArray(arg.del),
            makeDirs: toArray(arg.makeDirs),
        };
    });
}

export { cleanPreprocessor };
