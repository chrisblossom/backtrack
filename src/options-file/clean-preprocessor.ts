import path from 'path';
import { toArray } from '../utils/object-utils';
import { buildPath, rootPath } from '../config/paths';
import { Clean, NormalizedClean } from '../types';

interface Args {
	value?: Clean | Clean[];
}

function cleanPreprocessor({ value }: Args = {}): NormalizedClean[] {
	const allToArray = toArray(value);

	return allToArray.map((arg: Clean) => {
		const makeDirs = toArray(arg.makeDirs).map((makeDir) => {
			const resolvedMakeDir = path.resolve(buildPath, makeDir);
			return resolvedMakeDir;
		});

		const copy = toArray(arg.copy).map((copyArgs) => {
			const resolvedSrc = path.resolve(rootPath, copyArgs.src);
			const resolvedDest = path.resolve(buildPath, copyArgs.dest);

			const result = {
				src: resolvedSrc,
				dest: resolvedDest,
				hash: copyArgs.hash,
			};

			if (result.hash === undefined) {
				delete result.hash;
			}

			return result;
		});

		return {
			del: toArray(arg.del),
			makeDirs,
			copy,
		};
	});
}

export { cleanPreprocessor };
