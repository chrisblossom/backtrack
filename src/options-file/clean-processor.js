/* @flow */

import path from 'path';
import { buildPath } from '../config/paths';

import type { NormalizedClean } from '../types.js';

/**
 * Patterns are only allowed inside the buildPath.
 */
const base = {
    del: [],
    makeDirs: [],
};

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value: $ReadOnlyArray<NormalizedClean>,
    current?: NormalizedClean,
};

function cleanProcessor({ value, current = base }: Args) {
    return value.reduce(
        (acc, arg) => {
            arg.del.forEach((pattern) => {
                /**
                 * remove duplicates
                 */
                if (acc.del.includes(pattern) === false) {
                    acc.del.push(pattern);
                }
            });

            arg.makeDirs.forEach((dir) => {
                const resolvedDir = path.resolve(buildPath, dir);

                /**
                 * remove duplicates
                 */
                if (acc.makeDirs.includes(resolvedDir) === false) {
                    acc.makeDirs.push(resolvedDir);
                }
            });

            return acc;
        },
        {
            del: [...current.del],
            makeDirs: [...current.makeDirs],
        },
    );
}

export { cleanProcessor };
