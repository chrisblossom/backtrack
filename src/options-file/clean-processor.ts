import { NormalizedClean } from '../types';

/**
 * Patterns are only allowed inside the buildPath.
 */
const base = {
	del: [],
	makeDirs: [],
	copy: [],
};

interface Args {
	value: NormalizedClean[];
	current?: NormalizedClean;
}

function cleanProcessor({ value, current = base }: Args) {
	return value.reduce(
		(acc, arg) => {
			arg.del.forEach((pattern) => {
				/**
				 * remove duplicates
				 */
				if (acc.del.includes(pattern) === false) {
					acc.del.push(pattern);
				}
			});

			arg.makeDirs.forEach((dir) => {
				/**
				 * remove duplicates
				 */
				if (acc.makeDirs.includes(dir) === false) {
					acc.makeDirs.push(dir);
				}
			});

			arg.copy.forEach((copy) => {
				/**
				 * remove duplicates
				 */
				const isDuplicate = !!acc.copy.find((copyArg) => {
					return (
						copyArg.src === copy.src && copyArg.dest === copy.dest
					);
				});

				if (isDuplicate === false) {
					acc.copy.push(copy);
				}
			});

			return acc;
		},
		{
			del: [...current.del],
			makeDirs: [...current.makeDirs],
			copy: [...current.copy],
		},
	);
}

export { cleanProcessor };
