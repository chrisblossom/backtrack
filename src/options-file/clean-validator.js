/* @flow */

import { buildPath } from '../config/paths';
import { fileIsInsideDir } from '../utils/file-is-inside-dir';

import type { NormalizedClean } from '../types.js';

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value: $ReadOnlyArray<NormalizedClean>,
};

function cleanValidator({ value }: Args) {
    for (const arg of value) {
        for (const pattern of arg.del) {
            if (fileIsInsideDir(pattern, buildPath) === false) {
                throw new Error(
                    `del pattern '${pattern}' must be inside build directory`,
                );
            }
        }

        for (const dir of arg.makeDirs) {
            if (fileIsInsideDir(dir, buildPath) === false) {
                throw new Error(
                    `makeDirs '${dir}' must be inside build directory`,
                );
            }
        }
    }
}

export { cleanValidator };
