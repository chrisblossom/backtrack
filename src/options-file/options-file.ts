import { loadOptionsFile } from './load-options-file';
import { transformConfig } from './transform-config';
import { rootPath } from '../config/paths';

import { Lifecycles } from '../types';

function optionsFile(dirname: string = rootPath): Lifecycles {
    const baseConfig = loadOptionsFile(dirname);

    const config = transformConfig(baseConfig, dirname);

    return config;
}

export { optionsFile };
