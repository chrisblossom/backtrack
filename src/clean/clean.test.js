/* @flow */

import path from 'path';

require('./clean');
require('../options-file/clean-processor');
require('../options-file/clean-preprocessor');

const rawClean = (args) => require('./clean').clean(args);
const cleanParser = (args) =>
    require('../options-file/clean-processor').cleanProcessor(args);
const cleanPreprocessor = (args) =>
    require('../options-file/clean-preprocessor').cleanPreprocessor(args);

function clean(files) {
    const normalized = cleanPreprocessor({ value: files });
    const parsed = cleanParser({ value: normalized });

    return rawClean(parsed);
}

describe('clean', () => {
    const cwd = process.cwd();
    let del;
    let ensureDir;
    let copy;

    beforeEach(() => {
        jest.doMock('../utils/log', () => ({
            warn: jest.fn(),
            info: jest.fn(),
            error: jest.fn(),
            success: jest.fn(),
        }));

        jest.doMock('del', () =>
            jest.fn((pattern, options = {}) => {
                // $FlowIssue
                const delActual = require.requireActual('del');

                return delActual(pattern, {
                    ...options,
                    dryRun: true,
                });
            }),
        );
        // $FlowIssue
        del = require.requireMock('del');

        jest.doMock('fs-extra', () => ({
            ensureDir: jest.fn(() => Promise.resolve()),
            copy: jest.fn(() => Promise.resolve()),
        }));

        // $FlowIssue
        ensureDir = require.requireMock('fs-extra').ensureDir;
        // $FlowIssue
        copy = require.requireMock('fs-extra').copy;
    });

    afterEach(() => {
        process.chdir(cwd);
    });

    test('handles undefined', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/1/');
        process.chdir(dir);

        await rawClean();

        expect(del.mock.calls).toMatchSnapshot();
        expect(ensureDir.mock.calls).toEqual([]);
    });

    test('handles one clean', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/1/');
        process.chdir(dir);

        const options = {
            del: ['**/*'],
            makeDirs: ['nested/', 'another_nested/folder'],
            copy: [
                {
                    src: 'static',
                    dest: 'static',
                    hash: true,
                },
            ],
        };

        await clean(options);

        expect(del.mock.calls).toMatchSnapshot();
        expect(ensureDir.mock.calls).toMatchSnapshot();
        expect(copy.mock.calls).toMatchSnapshot();
    });

    test('handles multiple clean', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/1/');
        process.chdir(dir);

        const options = [
            {
                del: ['**/*', '!.git'],
                makeDirs: ['nested/', 'another_nested/folder'],
                copy: [
                    {
                        src: 'static',
                        dest: 'static',
                        hash: true,
                    },
                ],
            },
            {
                del: ['!two.js'],
            },
            {
                makeDirs: ['another_two_nested/folder'],
            },
            {
                copy: [
                    {
                        src: 'outside-build.js',
                        dest: 'static/outside-build.js',
                    },
                ],
            },
        ];

        const normalized = cleanPreprocessor({ value: options });
        const parsed = cleanParser({ value: normalized });

        await clean(parsed);

        expect(del.mock.calls).toMatchSnapshot();
        expect(ensureDir.mock.calls).toMatchSnapshot();
        expect(copy.mock.calls).toMatchSnapshot();
    });

    test('handles empty del', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/1/');
        process.chdir(dir);

        const options = [
            {
                makeDirs: ['nested/', 'another_nested/folder'],
            },
        ];

        await clean(options);

        expect(del.mock.calls).toMatchSnapshot();
        expect(ensureDir.mock.calls).toMatchSnapshot();
    });

    test('handles empty makeDirs', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/1/');
        process.chdir(dir);

        const options = [
            {
                del: ['**/*'],
            },
        ];

        await clean(options);

        expect(del.mock.calls).toMatchSnapshot();
        expect(ensureDir.mock.calls).toEqual([]);
    });

    test('merges duplicates in correct order', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/1/');
        process.chdir(dir);

        const options = [
            {
                del: ['duplicate/file', 'unique/file', 'duplicate/file'],
                makeDirs: ['duplicate/dir', 'unique/dir', 'duplicate/dir'],
            },
        ];

        await clean(options);

        expect(del.mock.calls).toMatchSnapshot();
        expect(ensureDir.mock.calls).toMatchSnapshot();
    });
});
