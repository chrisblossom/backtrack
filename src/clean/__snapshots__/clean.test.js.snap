// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clean handles empty del 1`] = `
Array [
  Array [
    Array [],
    Object {
      "cwd": "<PROJECT_ROOT>/dist",
      "dot": true,
    },
  ],
]
`;

exports[`clean handles empty del 2`] = `
Array [
  Array [
    "<PROJECT_ROOT>/dist/nested",
  ],
  Array [
    "<PROJECT_ROOT>/dist/another_nested/folder",
  ],
]
`;

exports[`clean handles empty del 3`] = `
Object {
  "del": Array [],
  "makeDirs": Array [
    "<PROJECT_ROOT>/dist/nested",
    "<PROJECT_ROOT>/dist/another_nested/folder",
  ],
}
`;

exports[`clean handles empty makeDirs 1`] = `
Array [
  Array [
    Array [
      "**/*",
    ],
    Object {
      "cwd": "<PROJECT_ROOT>/dist",
      "dot": true,
    },
  ],
]
`;

exports[`clean handles empty makeDirs 2`] = `
Object {
  "del": Array [
    "<PROJECT_ROOT>/dist/nested",
    "<PROJECT_ROOT>/dist/nested/nested.js",
    "<PROJECT_ROOT>/dist/one.js",
    "<PROJECT_ROOT>/dist/two.js",
  ],
  "makeDirs": Array [],
}
`;

exports[`clean handles multiple clean 1`] = `
Array [
  Array [
    Array [
      "**/*",
      "!.git",
      "!two.js",
    ],
    Object {
      "cwd": "<PROJECT_ROOT>/dist",
      "dot": true,
    },
  ],
]
`;

exports[`clean handles multiple clean 2`] = `
Array [
  Array [
    "<PROJECT_ROOT>/dist/nested",
  ],
  Array [
    "<PROJECT_ROOT>/dist/another_nested/folder",
  ],
  Array [
    "<PROJECT_ROOT>/dist/another_two_nested/folder",
  ],
]
`;

exports[`clean handles multiple clean 3`] = `
Object {
  "del": Array [
    "<PROJECT_ROOT>/dist/nested",
    "<PROJECT_ROOT>/dist/nested/nested.js",
    "<PROJECT_ROOT>/dist/one.js",
  ],
  "makeDirs": Array [
    "<PROJECT_ROOT>/dist/nested",
    "<PROJECT_ROOT>/dist/another_nested/folder",
    "<PROJECT_ROOT>/dist/another_two_nested/folder",
  ],
}
`;

exports[`clean handles one clean 1`] = `
Array [
  Array [
    Array [
      "**/*",
    ],
    Object {
      "cwd": "<PROJECT_ROOT>/dist",
      "dot": true,
    },
  ],
]
`;

exports[`clean handles one clean 2`] = `
Array [
  Array [
    "<PROJECT_ROOT>/dist/nested",
  ],
  Array [
    "<PROJECT_ROOT>/dist/another_nested/folder",
  ],
]
`;

exports[`clean handles one clean 3`] = `
Object {
  "del": Array [
    "<PROJECT_ROOT>/dist/nested",
    "<PROJECT_ROOT>/dist/nested/nested.js",
    "<PROJECT_ROOT>/dist/one.js",
    "<PROJECT_ROOT>/dist/two.js",
  ],
  "makeDirs": Array [
    "<PROJECT_ROOT>/dist/nested",
    "<PROJECT_ROOT>/dist/another_nested/folder",
  ],
}
`;

exports[`clean handles undefined 1`] = `
Array [
  Array [
    Array [],
    Object {
      "cwd": "<PROJECT_ROOT>/dist",
      "dot": true,
    },
  ],
]
`;

exports[`clean merges duplicates in correct order 1`] = `
Array [
  Array [
    Array [
      "duplicate/file",
      "unique/file",
    ],
    Object {
      "cwd": "<PROJECT_ROOT>/dist",
      "dot": true,
    },
  ],
]
`;

exports[`clean merges duplicates in correct order 2`] = `
Array [
  Array [
    "<PROJECT_ROOT>/dist/duplicate/dir",
  ],
  Array [
    "<PROJECT_ROOT>/dist/unique/dir",
  ],
]
`;

exports[`clean merges duplicates in correct order 3`] = `
Object {
  "del": Array [],
  "makeDirs": Array [
    "<PROJECT_ROOT>/dist/duplicate/dir",
    "<PROJECT_ROOT>/dist/unique/dir",
  ],
}
`;
